### PR TITLE
`MatrixSpace`: Support constructing `Hom(CombinatorialFreeModule)`

### PR DESCRIPTION
**Pull request by mkoeppe**
The `FreeModule` constructor was recently generalized so that it can create `CombinatorialFreeModule` objects. In particular,
```
sage: ZZ^3
Ambient free module of rank 3 over the principal ideal domain Integer Ring
sage: ZZ^['a','b','c']
Free module generated by {'a', 'b', 'c'} over Integer Ring
```

Here as a follow-up, we extend `MatrixSpace` in a similar way: When lists or enumerated sets of row indices and column indices are given, it constructs `CombinatorialFreeModule`s whose bases are indexed by these index sets, and then returns the homset. In particular,
```
sage: ZZ^(2, 3)
Full MatrixSpace of 2 by 3 dense matrices over Integer Ring
sage: ZZ^(['x','y'], ['a', 'b', 'c'])
Set of Morphisms from Free module generated by {'a', 'b', 'c'} over Integer Ring to Free module generated by {'x', 'y'} over Integer Ring in Category of finite dimensional modules with basis over Integer Ring
```

Follow-up PRs:
- extend `matrix` likewise; guiding application: adjacency / incidence "matrices" of graphs, where rows/columns are indexed by nodes and edges 
- see also #35801

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/matrix/matrix_space.py [hunk 1]
```diff
@@ -434,9 +434,75 @@ def get_matrix_class(R, nrows, ncols, sparse, implementation):
...
+    INPUT:
+
+    - ``base_ring`` -- a ring
+
+    - ``nrows`` or ``row_keys`` -- (nonnegative integer) the number of rows, or
+      a finite or enumerated family of arbitrary objects that index the rows
+      of the matrix
+
+    - ``ncols`` or ``column_keys`` -- (nonnegative integer, default ``nrows``)
+      the number of columns, or a finite or enumerated family of arbitrary objects
```
Comment (fchapoton) [inline src/sage/matrix/matrix_space.py]: I would prefer to remove "or enumerated" for safety.

Comment (mkoeppe) [reply src/sage/matrix/matrix_space.py]: OK, that's finite with me, done in 6885c2d

**Comment by mkoeppe** [comment]
Ready to go?

**Review by gmou3 — APPROVED** [review]

**Review by tscrim — COMMENTED** [review]
What happens if you mix the dimension and keys? For example, row keys with number of columns. (Of course, there is more evil input of row keys and row dimension, which might be good to check against with doctests.)

**Comment by mkoeppe** [comment]
It's checked for consistency, and errors are raised if there's a mismatch

**Comment by mkoeppe** [comment]
Yes, I can add more doctests for that. If there are specific ones that you'd like to see, let me know

**Comment by tscrim** [comment]
Just doctests testing the compatibility, such as
```
MatrixSpace(QQ, ['a','b'], 4)
MatrixSpace(QQ, 4, ['a','b'])
MatrixSpace(QQ, ['a','b'], ['x','y'], nrows=2)  # I think should pass
MatrixSpace(QQ, ['a','b'], ['x','y'], nrows=4)
MatrixSpace(QQ, ['a','b'], ['x','y'], ncols=2)  # I think should pass
MatrixSpace(QQ, ['a','b'], ['x','y'], ncols=4)
MatrixSpace(QQ, ['a','b'], ['x','y'], nrows=2, ncols=2)  # I think should pass
MatrixSpace(QQ, ['a','b'], ['x','y'], nrows=2, ncols=4)
MatrixSpace(QQ, ['a','b'], ['x','y'], nrows=4, ncols=4)
MatrixSpace(QQ, 4, ['a','b'], nrows=4, ncols=2)  # I think should pass if the second above passes
```

Also, your `INPUT:` block and docstring is not correct. There is `nrows_or_row_keys` (which I think should be changed to `row_keys`) and similarly for columns.

**Comment by mkoeppe** [comment]
> There is `nrows_or_row_keys` (which I think should be changed to `row_keys`)

It's intended as the name of a positional-only argument. 

> MatrixSpace(QQ, 4, ['a','b'], nrows=4, ncols=2)  # I think should pass if the second above passes

No, I disagree, it should be rejected as a case of "duplicate argument" nrows.

**Comment by tscrim** [comment]
> > There is `nrows_or_row_keys` (which I think should be changed to `row_keys`)
> 
> It's intended as the name of a positional-only argument.

Ah, sorry I missed it in the long list of arguments. Now there are more compatibility tests to be done as you have to mix with the `row_keys` and `nrows`.

> > MatrixSpace(QQ, 4, ['a','b'], nrows=4, ncols=2)  # I think should pass if the second above passes
> 
> No, I disagree, it should be rejected as a case of "duplicate argument" nrows.

It isn't a duplication per se. I could see it being used as a sanity check by a user. I would be a bit confused by this resulting in an error because of course it has 4 rows and 2 columns and the 4 is a shorthand for the index set [0, 1, 2, 3] in this setting.

Also, I believe this is true but I would like confirmation, the values that can be converted to `int` will always be less than `sys.maxsize`, correct?

**Comment by mkoeppe** [comment]
> the 4 is a shorthand for the index set [0, 1, 2, 3] in this setting.

Well, there's a clear distinction between passing 4 and passing range(4). 
One creates a FreeModule and the other creates a CombinatorialFreeModule.

**Comment by mkoeppe** [comment]
I have added the doctests (and input checking code), please take a look

**Comment by mkoeppe** [comment]
> the values that can be converted to `int` will always be less than `sys.maxsize`, correct?

No, `int(x)` converts any integer to an `int`. So we get, as expected, 
```
sage: MatrixSpace(ZZ, sys.maxsize+1)
OverflowError: number of rows and columns may be at most 9223372036854775807
```

**Comment by tscrim** [comment]
> > the values that can be converted to `int` will always be less than `sys.maxsize`, correct?
> 
> No, `int(x)` converts any integer to an `int`. So we get, as expected,
> 
> ```
> sage: MatrixSpace(ZZ, sys.maxsize+1)
> OverflowError: number of rows and columns may be at most 9223372036854775807
> ```

That didn't exactly answer my question. I don't know what limits Python has imposed on the `int` right now, and while it might work for our systems, it should be universal.

**Comment by mkoeppe** [comment]
> > > the 4 is a shorthand for the index set [0, 1, 2, 3] in this setting.
> > 
> > 
> > Well, there's a clear distinction between passing 4 and passing range(4). One creates a FreeModule and the other creates a CombinatorialFreeModule.
> 
> If _both_ of the arguments were integers, then I would agree. However, because one of the inputs is a list, passing `4` gets implicitly converted into a list before being passed along. 

No, there's no such mechanism. It constructs a morphism between a FreeModule and a CombinatorialFreeModule in this case.
```
      sage: MatrixSpace(QQ, ['a','b'], 4)
        Set of Morphisms (Linear Transformations)
         from Vector space of dimension 4 over Rational Field
           to Free module generated by {'a', 'b'} over Rational Field
```

**Comment by mkoeppe** [comment]
> > > the values that can be converted to `int` will always be less than `sys.maxsize`, correct?
> > 
> > 
> > No, `int(x)` converts any integer to an `int`. So we get, as expected,
> > ```
> > sage: MatrixSpace(ZZ, sys.maxsize+1)
> > OverflowError: number of rows and columns may be at most 9223372036854775807
> > ```
> 
> That didn't exactly answer my question. I don't know what limits Python has imposed on the `int` right now, and while it might work for our systems, it should be universal.

I don't understand what you mean. What limits are you talking about?

**Comment by tscrim** [comment]
Okay, I've checked, in Python3, `int` is bounded by the size of the memory, which should always be larger than `sys.maxsize` (unless your machine is completely crazy).

**Comment by tscrim** [comment]
> > > > the 4 is a shorthand for the index set [0, 1, 2, 3] in this setting.
> > > 
> > > 
> > > Well, there's a clear distinction between passing 4 and passing range(4). One creates a FreeModule and the other creates a CombinatorialFreeModule.
> > 
> > 
> > If _both_ of the arguments were integers, then I would agree. However, because one of the inputs is a list, passing `4` gets implicitly converted into a list before being passed along.
> 
> No, there's no such mechanism. It constructs a morphism between a `FreeModule` and a `CombinatorialFreeModule` in this case.
> 
> ```
>       sage: MatrixSpace(QQ, ['a','b'], 4)
>         Set of Morphisms (Linear Transformations)
>          from Vector space of dimension 4 over Rational Field
>            to Free module generated by {'a', 'b'} over Rational Field
> ```
> 
> Edit: I'll add that many of these homsets are not fully implemented, but that's outside of the scope of this PR. I have added some `# known bug` now to document this. For completing the implementation, see for example:
> 
>     * [Fix scaling of vector space morphisms on the left #37709](https://github.com/sagemath/sage/pull/37709)
> 
>     * [Make homsets of FiniteDimensionalModulesWithBasis finite-dimensional #37516](https://github.com/sagemath/sage/pull/37516)

Ah I see. This is acceptable then.

Although now I notice another issue. I don't quite like how, e.g., setting `ncols` overrides the CFM to a standard free module:
````
sage: MatrixSpace(QQ, ['a','b'], ['x','y'], ncols=2)
Set of Morphisms (Linear Transformations)
 from Vector space of dimension 2 over Rational Field
  to Free module generated by {'a', 'b'} over Rational Field
```
I am treating the `ncols` as verification here of the dimension.

**Comment by mkoeppe** [comment]
> Although now I notice another issue. I don't quite like how, e.g., setting `ncols` overrides the CFM to a standard free module:

That's a bug.

**Comment by mkoeppe** [comment]
Fixed now

**Comment by mkoeppe** [comment]
Any more comments or can we merge it?

**Review by tscrim — APPROVED** [review]
Yes, this is acceptable to me.

**Comment by mkoeppe** [comment]
Thanks!

**Pull request merged**